halo2-base/src/gates/tests/prop_test.rs

```diff
@@ -0,0 +1,333 @@
+    }
+}
+
+// Ground truth functions for Property Testing
+
+fn add_ground_truth<F: ScalarField>(inputs: &[QuantumCell<F>]) -> F {
+    *inputs[0].value() + *inputs[1].value()
+}
+
+fn sub_ground_truth<F: ScalarField>(inputs: &[QuantumCell<F>]) -> F {
+    *inputs[0].value() - *inputs[1].value()
+}
+
+fn neg_ground_truth<F: ScalarField>(input: QuantumCell<F>) -> F {
+    -(*input.value())
+}
+
+fn mul_ground_truth<F: ScalarField>(inputs: &[QuantumCell<F>]) -> F {
+    *inputs[0].value() * *inputs[1].value()
+}
+
+fn mul_add_ground_truth<F: ScalarField>(inputs: &[QuantumCell<F>]) -> F {
+    *inputs[0].value() * *inputs[1].value() + *inputs[2].value()
+}
+
+fn mul_not_ground_truth<F: ScalarField>(inputs: &[QuantumCell<F>]) -> F {
+    *inputs[0].value() * (F::one() - *inputs[1].value())
+}
+
+fn div_unsafe_ground_truth<F: ScalarField>(inputs: &[QuantumCell<F>]) -> F {
+    inputs[0].value().invert().unwrap() * *inputs[1].value()
+}
+
+fn inner_product_ground_truth<F: ScalarField>(inputs: &(Vec<QuantumCell<F>>, Vec<QuantumCell<F>>)) -> F {
+    inputs.0.iter().zip(inputs.1.iter()).fold(F::zero(),|acc, (a, b)| acc + (*a.value() * *b.value()))    
+}
+
+fn inner_product_left_last_ground_truth<F: ScalarField>(inputs: &(Vec<QuantumCell<F>>, Vec<QuantumCell<F>>)) -> (F, F) {
+    let product = inner_product_ground_truth(inputs);
+    let last = *inputs.0.last().unwrap().value();
+    (product, last)
+}
+
+fn inner_product_with_sums_ground_truth<F: ScalarField>(input: &(Vec<QuantumCell<F>>, Vec<QuantumCell<F>>)) -> Vec<F> {
+    let (a, b) = &input;
+    let mut result = Vec::new();
+    let mut sum = F::zero();
+    // TODO: convert to fold
+    for (ai, bi) in a.iter().zip(b) {
+        let product = *ai.value() * *bi.value();
+        sum += product;
+        result.push(sum);
+    }
+    result
+}
+
+fn sum_products_with_coeff_and_var_ground_truth<F: ScalarField>(input: &(Vec<(F, QuantumCell<F>, QuantumCell<F>)>, QuantumCell<F>)) -> F {
+    let expected = input.0.iter().fold(F::zero(), |acc, (coeff, cell1, cell2)| {
+        acc + *coeff * *cell1.value() * *cell2.value()
+    }) + *input.1.value();
+    expected
+}
+
+fn and_ground_truth<F: ScalarField>(inputs: &[QuantumCell<F>]) -> F {
+    *inputs[0].value() * *inputs[1].value()
+}
+
+fn not_ground_truth<F: ScalarField>(a: &QuantumCell<F>) -> F {
+    F::one() - *a.value()
+}
+
+fn select_ground_truth<F: ScalarField>(inputs: &[QuantumCell<F>]) -> F {
+    (*inputs[0].value() *  *inputs[2].value()) +  (*inputs[1].value() *( *inputs[2].value() - F::one()))
+}
+
+fn or_and_ground_truth<F: ScalarField>(inputs: &[QuantumCell<F>]) -> F {
+    let bc_val = *inputs[1].value() * inputs[2].value();
+    bc_val + inputs[0].value() - bc_val * inputs[0].value()
+}
+
+fn idx_to_indicator_ground_truth<F: ScalarField>(inputs: (QuantumCell<F>, usize)) -> Vec<F> {
+    let (idx, size) = inputs;
+    let mut indicator = vec![F::zero(); size];
+    let mut idx_value = size + 1;
+    for i in 0..u64::MAX {
```

Comment:
This takes forever if it's looping to `2^64 - 1`.

Should only be for looping `i in 0..size`. Or when the `idx` is generated it should only be generated within range `0..size`, otherwise most test cases are going to give `0` indicator